File: src/main/java/it/alian/gun/mesmerize/compat/ShieldBlocking.java
package it.alian.gun.mesmerize.compat;

import org.bukkit.Material;
import org.bukkit.entity.Player;

public abstract class ShieldBlocking {

    private static ShieldBlocking impl;

    public static void init() {
        try {
            Material.valueOf("SHIELD");
            impl = new Impl_1_9();
        } catch (Throwable t) {
            impl = new AbstractImpl();
        }
    }

    public static boolean check(Player p) {
        return impl.checkShield(p);
    }

    public abstract boolean checkShield(Player player);

    private static class Impl_1_9 extends ShieldBlocking {

        @Override
        public boolean checkShield(Player player) {
            return player.getEquipment().getItemInMainHand().getType() == Material.SHIELD && player.isBlocking();
        }

    }

    private static class AbstractImpl extends ShieldBlocking {

        @Override
        public boolean checkShield(Player player) {
            return false;
        }
    }

}

File: src/main/scala/it/alian/gun/mesmerize/compat/ShieldBlocking.scala
package it.alian.gun.mesmerize.compat

import it.alian.gun.mesmerize.scalaapi.Prelude._
import org.bukkit.Material
import org.bukkit.entity.Player

object ShieldBlocking extends (Player => Boolean) {
  private val impl = try {
    Material.valueOf("SHIELD")
    new Impl_1_9
  } catch {
    case _: Throwable => new AbstractImpl
  }

  /**
    * @param player the attacker
    * @return true if should ignore shield blocking
    */
  def apply(player: Player): Boolean = impl(player)

  private class Impl_1_9 extends (Player => Boolean) {
    override def apply(player: Player): Boolean = ((player.getEquipment.getItemInMainHand.getType ne Material.SHIELD)
      || !player.isBlocking) && config("general.ignoreShieldBlocking")
  }

  private class AbstractImpl extends (Player => Boolean) {
    override def apply(player: Player) = false
  }

}
